src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java

```diff
@@ -0,0 +1,950 @@
+        logger.debug("{}: garbage: {}/{}", tableName, globalGarbage, globalSizeOnDisk);
+
+        // look for garbage cleanup.
+        List<OverlappingSet> overlappingTombstones = getDistinctOverlappingSSTables(categorizedSSTables.expiredTombstones);
+
+        Stream<DataCompactionCandidate> candidateStream = overlappingTombstones.stream()
+                .map(ot -> DataCompactionCandidate.get(categorizedSSTables.data, ot));
+
+        // if we are crossing the global threshold, just pick up the set with the most garbage, otherwise only choose the sets that are full of garbage.
+        // This is done so that many localized deletes can result in compaction.
+        candidateStream = isTooMuchGarbage(globalSizeOnDisk, globalGarbage, true)
+                ? candidateStream
+                : candidateStream.filter(c -> isTooMuchGarbage(c.stats.onDiskLength, c.estimatedGarbage, false));
+        Optional<DataCompactionCandidate> candidate = candidateStream.max(Comparator.comparingDouble(c -> c.estimatedGarbage));
+
+        if (!candidate.isPresent())
+        {
+            logger.debug("{}: no significant garbage found", tableName);
+            return Optional.empty();
+        }
+
+        List<SSTableStats> maxGarbageData = candidate.get().data;
+        OverlappingSet maxGarbageTombstones = candidate.get().tombstones;
+
+        TimeWindow dataSSTableTimeWindowMs = TimeWindow.merge(maxGarbageData.stream().map(s -> s.timeWindowMs).collect(Collectors.toList()));
+        if (logger.isDebugEnabled())
+        {
+            logger.debug("{}: candidate: {}, data_tw: {}", tableName, candidate.get(), dataSSTableTimeWindowMs);
+        }
+
+        // If there are too many files OR (data size is over the threshold and the data set is too wide),
+        // we need a sure shot way of reducing them.
+        if (maxGarbageTombstones.sstables.size() + maxGarbageData.size() > getMaxFileCountForCompaction() || dataSSTableTimeWindowMs.duration > 2 * getWindowSizeInMs())
+        {
+            if (maxGarbageTombstones.timeWindowMs.duration > 2 * getWindowSizeInMs() || maxGarbageTombstones.sstables.size() > 1)
+            {
+                logger.debug("{}: too many tombstone files", tableName);
+                return Optional.of(new SSTablesForCompaction(toSSTableReader(maxGarbageTombstones.sstables), true, true));
+            }
+
+            List<OverlappingSet> overlappingDataSets = getDistinctOverlappingSSTables(maxGarbageData);
+            long windowSzInMs = getWindowSizeInMs();
+
+            logger.debug("{}: too many data files", tableName);
+            // we are here, which means that tombstone.timerange <= 2 * window && tombstone.size() <= 1, and there could be too many data files.
+            // data files could be too wide or many small files.
+            // we would like to merge all those files that are overlapping or are in a single compaction window.
+            return overlappingDataSets.stream()
```

Comment:
Consider refactoring this, feels a bit hard to follow what's happening, especially with the multiple single character temp vars (s, v, l) and the types (we're dealing with Map of List of Overlapping Set of SSTable from what I can see). Might be better to just replace with basic for loops.

Reply:
Also I'd invert the if condition. 
```
Instead of 
find candidate
if(candidate too large){
do different stuff
} else{
return candidate
}

maybe do:

if(candidate not too large){
return candidate
}
do different stuff
```

Reply:
You have correct understanding.
Normal loops were introducing some really large types. streams looked a bit clear.
Yes the method in itself is quite large at the moment and can be split.